estransport/connection.go

```diff
@@ -0,0 +1,300 @@
+			cp.live = append(cp.live, c)
+			c.Unlock()
+
+			return c, nil
+		}
+		return nil, errors.New("no connection available")
+	}
+
+	if cp.curr >= len(cp.live) {
+		cp.curr = len(cp.live) - 1
+	}
+
+	if cp.curr < 0 {
+		return nil, errors.New("no connection available")
+	}
+
+	c = cp.live[cp.curr]
+	cp.curr = (cp.curr + 1) % len(cp.live)
+
+	return c, nil
+}
+
+// Remove removes a connection from the pool.
+//
+func (cp *roundRobinConnectionPool) Remove(c *Connection) error {
+	c.Lock()
+
+	if c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already removed %s\n", c.URL)
+		}
+		c.Unlock()
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Removing %s...\n", c.URL)
+	}
+	c.markAsDead()
+	cp.scheduleResurrect(c)
+	c.Unlock()
+
+	cp.Lock()
+	defer cp.Unlock()
+
+	// Push item to dead list and sort slice by number of failures
+	cp.dead = append(cp.dead, c)
+	sort.Slice(cp.dead, func(i, j int) bool {
+		c1 := cp.dead[i]
+		c2 := cp.dead[j]
+		c1.Lock()
+		if c1 != c2 {
+			c2.Lock()
+		}
+		defer c1.Unlock()
+		if c1 != c2 {
+			defer c2.Unlock()
+		}
+
+		res := c1.Failures > c2.Failures
+		return res
+	})
+
+	// Check if connection exists in the list. Return nil if it doesn't,
+	// because another goroutine might have already removed it.
+	index := -1
+	for i, conn := range cp.live {
+		if conn == c {
+			index = i
+		}
+	}
+	if index < 0 {
+		return nil
+	}
+
+	// Remove item; https://github.com/golang/go/wiki/SliceTricks
+	copy(cp.live[index:], cp.live[index+1:])
+	cp.live = cp.live[:len(cp.live)-1]
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.dead = cp.dead
+		cp.metrics.live = cp.live
+		cp.metrics.Unlock()
+	}
+
+	return nil
+}
+
+// Resurrect removes the connection from the dead list and adds it to the pool.
+//
+// TODO(karmi): Add a pluggable strategy as argument, eg. "optimistic", "ping".
+//
+func (cp *roundRobinConnectionPool) Resurrect(c *Connection) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	c.Lock()
+	defer c.Unlock()
+
+	if !c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already resurrected %s\n", c.URL)
+		}
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Resurrecting %s\n", c.URL)
+	}
+
+	c.markAsLive()
+
+	cp.live = append(cp.live, c)
+	index := -1
+	for i, conn := range cp.dead {
+		if conn == c {
+			index = i
+		}
+	}
+	if index >= 0 {
+		// Remove item; https://github.com/golang/go/wiki/SliceTricks
+		copy(cp.dead[index:], cp.dead[index+1:])
+		cp.dead = cp.dead[:len(cp.dead)-1]
+	}
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.dead = cp.dead
+		cp.metrics.live = cp.live
+		cp.metrics.Unlock()
+	}
+
+	return nil
+}
+
+// scheduleResurrect schedules the connection to be resurrected.
+//
+func (cp *roundRobinConnectionPool) scheduleResurrect(c *Connection) {
+	factor := func(a, b int) float64 {
+		if a > b {
+			return float64(b)
+		}
+		return float64(a)
+	}(c.Failures-1, defaultResurrectTimeoutFactorCutoff)
```

Comment:
`factor := math.Min(float64(c.Failures-1), defaultResurrectTimeoutFactorCutoff)` ?

Reply:
Argh, embarrasing :) Fixed in f7cbe9a.